### PR TITLE
Set the server sig algs before calling the session_secret_cb 

### PR DESCRIPTION
**Pull request by mattcaswell**
Setting the server sig algs sets up the certificate "s3->tmp.valid_flags".
These are needed when calling ssl3_choose_cipher() which can happen
immediately after calling the session_secret_cb

Fixes https://github.com/openssl/openssl/issues/24213

The sessions_secret_cb was undocumented and untested - so we also add a test case and add a man page for it.

### CODE REVIEW

**Review by nhorman — APPROVED** [review]

**Comment by mattcaswell** [comment]
> CI is relevant

Fixed (hopefully)

**Review by nhorman — APPROVED** [review]

**Review by beldmit — APPROVED** [review]
LGTM

File: doc/man3/SSL_set_session_secret_cb.pod [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+On the server side the set of ciphersuites offered by the peer is provided in
+the B<peer_ciphers> stack. Optionally the callback may select the preferred
+ciphersuite by setting it in B<*cipher>.
+
+On the client side the B<peer_ciphers> stack will always be NULL. The callback
+may specify the preferred cipher in B<*cipher> and this will be associated with
+the SSL_SESSION - but it does not affect the ciphersuite selected by the server.
+
+The callback is also supplied with an additional argument in B<arg> which is the
+argument that was provided to the original SSL_set_session_secret_cb() call.
```
Comment (t8m) [inline doc/man3/SSL_set_session_secret_cb.pod:44]: Please use `I<argument>` and `B<SSL_SESSION>` in this text.

**Review by t8m — CHANGES_REQUESTED** [review]
Just one doc style nit.

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
@beldmit @nhorman please reconfirm

**Comment by beldmit** [comment]
My approval persists

**Comment by nhorman** [comment]
Approval holds

**Comment by t8m** [comment]
Merged to the master, 3.3 and 3.2 branches. Thank you.

**Pull request closed without merging**